pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -3163,6 +3165,7 @@ public String startReplaceSegments(String tableNameWithType, List<String> segmen
           if (!segmentsToCleanUp.isEmpty()) {
             LOGGER.info("Cleaning up the segments while startReplaceSegments: {}", segmentsToCleanUp);
             deleteSegments(tableNameWithType, segmentsToCleanUp);
+            waitForSegmentsToDelete(tableNameWithType, segmentsToCleanUp, SEGMENT_CLEANUP_TIMEOUT_MS);
```

Comment:
is waitForSegmentsToDelete required for correctness or more for improvement purpose? If latter, perhaps add a feature switch to turn it on/off.

Reply:
This is more for correctness. By doing so, it guarantees at most 2 snapshots are stored and we will not violate the storage quota. It also helps the race condition mentioned in the description. The client side may need to change the socket timeout based on this change.

Reply:
@klsince Essentially, because of async deletion, there can be a short duration where 3 data snapshots (old, current, new) exist in the table and this can fill up the disk space pretty easily. To avoid that, we would want to make sure to finish "clean up" before we allow the new data push.

@jtao15 I actually think that we should consider making this optional. For instance, S3 file system implementation needs to copy data and then delete the previous one for `doMove()`. In this case, it will take a long time to delete segments.

Reply:
@klsince I discussed with @snleee offline:
1. The deletion only happens when `forceCleanup=true`
2. Controller will fail the request to start the protocol after 20 mins, but the deletion will keep going. For the file system that takes long time to delete, the client side will finally succeed after some retries.

Based on above, we should be good to change the deletion to blocking.